src/Microsoft.AspNetCore.Razor.Evolution/DirectiveTokenDescriptorComparer.cs

```diff
@@ -0,0 +1,44 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using Microsoft.Extensions.Internal;
+
+namespace Microsoft.AspNetCore.Razor.Evolution
+{
+    internal class DirectiveTokenDescriptorComparer : IEqualityComparer<DirectiveTokenDescriptor>
+    {
+        public static readonly DirectiveTokenDescriptorComparer Default = new DirectiveTokenDescriptorComparer();
+
+        protected DirectiveTokenDescriptorComparer()
+        {
+        }
+
+        public bool Equals(DirectiveTokenDescriptor descriptorX, DirectiveTokenDescriptor descriptorY)
+        {
+            if (descriptorX == descriptorY)
+            {
+                return true;
+            }
+
+            return descriptorX != null &&
+                string.Equals(descriptorX.Value, descriptorY.Value, StringComparison.Ordinal) &&
```

Comment:
If descriptorX != null then descriptorY != null because the == statement above would have short-circuited.

What you're suggesting is that we return `false` instead of `true` if both the descriptors being compared are `null`. Probably smart. We should do this for our other comparers because they fall into the same trap.